### PR TITLE
Add ability to load measured attenuation profile for material

### LINKED ISSUE #28139: Add ability to import externally generated attenuation profile into absorption calculation
For environment components made out of crystalline materials the attenuation effect may consist of absorption plus a scattering effect containing some Bragg edges. The absorption calculation (which should perhaps be renamed attenuation calculation) should be enhanced to allow an externally measured attenuation profile against wavelength to be imported into the calculation
This kind of profile has been generated by Pearl for their sintered diamond anvils. It is available as a csv format file

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

Add feature where an externally generated attenuation profile can be loaded into the material definition from a .DAT text file. This will override the normal exp(-alpha*thickness) calculation of the attenuation. The new feature was requested as part of the absorption calculation for the Pearl instrument where some parts of the environment are crystalline

Found the lookup\interpolation on the loaded profile was slowing down the absorption
Monte Carlo calculation. So improved performance of Kernel::Interpolation. Changed it to use vector of pairs instead of two separate vectors to reduce amount of navigation around vectors

Considered adding feature into AttenuationProfile where I add any interpolated points into the override attenuation profile to prevent interpolation next iteration of Monte Carlo simulation. However wasn't convinced this was going to be thread-safe during the parallel part of the absorption calculation - may revisit this. While doing this investigation I fixed the following issues with the function Interpolation::AddPoint:

a) it doesn't add the same point twice
b) it reuses the lower_bound function rather than stepping through the vector sequentially

I removed the exception from findIndexOfNextLargerValue because the end() iterator is a valid result if the x value supplied equals the last x value already in the vector. Could have left check in for begin() but didn't seem much point

Finally - the interpolation outside the range is not v good. The data at the extremes is noisy so we end up extrapolating based on random gradient in extreme 2 points. Need to either fit or use more points to determine gradient on future PR

**To test:**

Run the following Python script to load the Pearl sample environment details which now include an externally measured .DAT file containing the attenuation profile for the diamond Anvil in the Pearl environment. It then runs an absorption calculation on this environment specification with a small number of scenarios (100):
```
from __future__ import print_function, division, unicode_literals

from mantid.simpleapi import (ConvertUnits, Load, LoadSampleShape, MonteCarloAbsorption, SetSample, GroupDetectors)
import matplotlib.pyplot as plt

# Grab some data. This happens to be a run with a 10mm vanadium sphere

# select spectra for detector ids 1040, 2040, 3040, 4040, 5040, 6040, 7040, 8040, 9040
sample_env_mesh = Load(Filename='PEARL00073987.raw', SpectrumList='48,126,204,282,360,438,516,594,672')
# MonteCarloAbsorption works in wavelength
sample_env_mesh = ConvertUnits(InputWorkspace=sample_env_mesh, Target='Wavelength')

SetSample(InputWorkspace=sample_env_mesh, Environment={'Name':'Pearl'},Material={'ChemicalFormula': 'Ni'})

sample_env_mesh_abscorr = MonteCarloAbsorption(InputWorkspace=sample_env_mesh, EventsPerPoint=100)
```
Further tests could include editing the attenuation file path that is in the file instrument\sampleenvironments\ISIS\Pearl.xml (see materials section at top)

Fixes #28139 .

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by cailafinn — CHANGES_REQUESTED** [review]
I can't do the functional testing at the moment due to some technical issues, but I do have a few code comments, mostly about style.

File: Framework/DataHandling/src/ReadMaterial.cpp [hunk 1]
```diff
@@ -34,7 +34,8 @@ ReadMaterial::validateInputs(const MaterialParameters &params) {
                                      "no ChemicalFormula or AtomicNumber is "
                                      "given.";
     }
-    if (isEmpty(params.attenuationXSection)) {
+    if ((isEmpty(params.attenuationXSection) &&
+         (params.attenuationProfileFileName.empty()))) {
```
[the remark this answers is not in the document]
Comment (DannyHindson) [reply Framework/DataHandling/src/ReadMaterial.cpp]: Yes seem to have got a bit carried away there!

File: Framework/DataHandling/test/LoadSampleEnvironmentTest.h [hunk 2]
```diff
@@ -35,9 +35,6 @@ class LoadSampleEnvironmentTest : public CxxTest::TestSuite {
     LoadSampleEnvironment alg;
     TS_ASSERT_THROWS_NOTHING(alg.initialize());
     TS_ASSERT(alg.isInitialized());
-
-    TSM_ASSERT_EQUALS("should be 23 properties here", 23,
```
[the remark this answers is not in the document]
Comment (DannyHindson) [reply Framework/DataHandling/test/LoadSampleEnvironmentTest.h]: I didn't feel that counting the number of parameters was a useful, functional test. Also aside from my own personal view on this it's not a test that's performed on the majority of algorithms so I felt it should be removed for consistency also

File: Framework/Kernel/inc/MantidKernel/Interpolation.h [hunk 3]
```diff
@@ -74,7 +76,7 @@ class MANTID_KERNEL_DLL Interpolation {
   Unit_sptr getYUnit() const { return m_yUnit; };
 
   /// return false if no data has been added
-  bool containData() const { return !m_x.empty() ? true : false; }
+  bool containData() const { return !m_data.empty() ? true : false; }
```
[the remark this answers is not in the document]
Comment (DannyHindson) [reply Framework/Kernel/inc/MantidKernel/Interpolation.h]: Having stared at it for a while I'm not seeing a good reason. I'll simplify as you've suggested

**Review by cailafinn — DISMISSED** [review]
Functions well and code looks good. Nice work :+1:

File: Framework/DataHandling/src/SetSampleMaterial.cpp [hunk 4]
```diff
@@ -74,6 +75,11 @@ void SetSampleMaterial::init() {
                   "Optional:  This total scattering cross-section (coherent + "
                   "incoherent) for the sample material in barns will be used "
                   "instead of tabulated");
+  const std::vector<std::string> extensions{".DAT"};
+  declareProperty(
```
Comment (martyngigg) [inline Framework/DataHandling/src/SetSampleMaterial.cpp]: We unfortunately need to be careful with changes in the order of algorithm properties. The Python `simpleapi` functions auto generate their signatures and if we change the order then any calls in Python scripts that use only positional parameters will break. We have no other option than to put new things at the end.

File: Framework/Kernel/inc/MantidKernel/AttenuationProfile.h [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#include "MantidKernel/DllConfig.h"
+#include "MantidKernel/Interpolation.h"
+#include <vector>
+
+namespace Mantid {
+namespace Kernel {
+
+class MANTID_KERNEL_DLL AttenuationProfile {
+public:
+  AttenuationProfile(const std::string inputFileName,
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/AttenuationProfile.h]: For consistency with good practice can we use `const std::string &`?

File: Framework/Kernel/src/Material.cpp [hunk 6]
```diff
@@ -180,6 +180,11 @@ void Material::calculateTotalScatterXSection() {
     m_totalScatterXSection = weightedTotal;
   }
 }
+
+void Material::setAttenuationProfile(AttenuationProfile attenuationOverride) {
+  m_attenuationOverride = attenuationOverride;
```
Comment (martyngigg) [inline Framework/Kernel/src/Material.cpp]: To espouse good practice can we do `m_attenuationOverride = std::move(attenuationOverride);`

File: Framework/Kernel/src/MaterialBuilder.cpp [hunk 7]
```diff
@@ -210,6 +210,27 @@ MaterialBuilder &MaterialBuilder::setAbsorptionXSection(double xsec) {
...
+  }
+  return *this;
+}
+
+/**
+ * Set a value for the attenuation profile search path
+ * @param path Path to search
+ */
+void MaterialBuilder::setAttenuationSearchPath(std::string path) {
+  m_attenuationFileSearchPath = path;
```
Comment (martyngigg) [inline Framework/Kernel/src/MaterialBuilder.cpp]: `m_attenuationFileSearchPath = std::move(path);?`

File: Framework/Kernel/inc/MantidKernel/MaterialXMLParser.h [hunk 8]
```diff
@@ -27,7 +27,7 @@ class MANTID_KERNEL_DLL MaterialXMLParser {
   static constexpr const char *MATERIAL_TAG = "material";
 
   Material parse(std::istream &istr) const;
-  Material parse(Poco::XML::Element *node) const;
+  Material parse(Poco::XML::Element *node, std::string XMLFilePath = "") const;
```
Comment (martyngigg) [inline Framework/Kernel/inc/MantidKernel/MaterialXMLParser.h]: `const std::string &` ?

File: Framework/DataHandling/src/LoadSampleEnvironment.cpp [hunk 9]
```diff
@@ -114,6 +114,11 @@ void LoadSampleEnvironment::init() {
                   "Optional:  This total scattering cross-section (coherent + "
                   "incoherent) for the sample material in barns will be used "
                   "instead of tabulated");
+  const std::vector<std::string> attExtensions{".DAT"};
```
Comment (martyngigg) [inline Framework/DataHandling/src/LoadSampleEnvironment.cpp]: Same comment as with `SetSampleMaterial`.

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/DataHandling/src/SetSampleMaterial.cpp (hunk 4 above)
Comment (DannyHindson) [reply Framework/DataHandling/src/SetSampleMaterial.cpp]: I had paused to consider this - the validations on this algorithm (and LoadSampleEnvironment) means they can't be called using positional arguments so I decided it was safe to inserted the new parameter in the logical place next to the absorption cross section. I don't think there can be any scripts out there that will be broken by this. What do you think?

Comment (martyngigg) [reply Framework/DataHandling/src/SetSampleMaterial.cpp]: Yes I see what you are saying. It's not possible to form an ordered list of the parameters that is valid so they must have been specified by keyword. I agree this is fine.

Sorry for the confusion. I didn't think to check for `validateInputs` and made my usual comment about when I see changes like this.

Comment (DannyHindson) [reply Framework/DataHandling/src/SetSampleMaterial.cpp]: No problem, thanks for have a good look through

**Review by martyngigg — APPROVED** [review]

**Pull request merged**